### PR TITLE
Update ESLint rules and setup GitHub Actions

### PR DESCRIPTION
**Pull request by Sea-n**
This PR added 8 ESLint rules as errors and fixed existing codes.  
The following is the bad and good example for each rule.

# Errors
Mostly can be auto-fixes. Will cause CI to fails (and should be blocking merge process).
## [object-curly-spacing](https://eslint.org/docs/rules/object-curly-spacing)
```js
var obj = { 'foo': 'bar' };
var obj = {'foo': 'bar'};
```
## [space-before-blocks](https://eslint.org/docs/rules/space-before-blocks)
```js
if (cond){
}
if (cond) {
}
```
## [no-trailing-spaces](https://eslint.org/docs/rules/no-trailing-spaces)
```js
var foo;    
var foo;
```
## [no-unsafe-negation](https://eslint.org/docs/rules/no-unsafe-negation)
```js
if (!method in API)
if (!(method in API))
```
## [keyword-spacing](https://eslint.org/docs/rules/keyword-spacing)
```js
if(cond) {
}
if (cond) {
}

```
## [spaced-comment](https://eslint.org/docs/rules/spaced-comment)
```js
//Hello
// Hello
```
## [comma-spacing](https://eslint.org/docs/rules/comma-spacing)
```js
var foo = [1,2,3];
var foo = [1, 2, 3];
```
## [brace-style](https://eslint.org/docs/rules/brace-style)
```js
if (cond) {
  foo();
}
else {
  bar();
}

if (cond) {
  foo();
} else {
  bar();
}

```

# Warnings
Should be complied with. Will not cause CI to fail.
- [no-unused-vars](https://eslint.org/docs/rules/no-unused-vars)
- [camelcase](https://eslint.org/docs/rules/camelcase)
- [max-len](https://eslint.org/docs/rules/max-len) (120 chars)

# Off
Consider adding the following rules after discussion.
- [no-extra-semi](https://eslint.org/docs/rules/no-extra-semi)
- [comma-dangle](https://eslint.org/docs/rules/comma-dangle)
- [indent](https://eslint.org/docs/rules/indent)
- [quotes](https://eslint.org/docs/rules/quotes)
- [semi](https://eslint.org/docs/rules/semi)

### CODE REVIEW

**Comment by Sea-n** [comment]
\* Please don't merge this PR until the next meeting, thanks 🙇‍♂️

File: .github/workflows/eslint.yml [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+on: [push, pull_request]
+jobs:
+  lint:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - name: Install modules
+        run: yarn
+      - name: Run ESLint
+        run: yarn run eslint . --ext .mjs
```
Comment (ktpss95112) [inline .github/workflows/eslint.yml:11]: also `.js` files?

File: common/extlib/client-extension-manager.mjs [hunk 2]
```diff
@@ -121,7 +120,7 @@ class ClientExtensionManager {
     if (!typeof methodName === 'string') {
```
Comment (ktpss95112) [inline common/extlib/client-extension-manager.mjs:120]: should be `(typeof methodName !== 'string')`?

Comment (ktpss95112) [reply common/extlib/client-extension-manager.mjs:120]: Is there any eslint rule that will catch this type of logic error?

**Comment by Sea-n** [comment]
目前用自訂 rules 並修復完成後，共剩下 90 個 warning、0 個 error

如果加上 `eslint:recommended` 規則集，會多出這 14 個 error，短時間內我無法獨立修復
```
hitcon-online/common/extlib/client-extension-manager.mjs
  117:9  error  Unexpected constant condition  no-constant-condition
  120:9  error  Unexpected constant condition  no-constant-condition
  144:9  error  Unexpected constant condition  no-constant-condition
  158:9  error  Unexpected constant condition  no-constant-condition

hitcon-online/common/rpc-directory/data-store.mjs
  186:12 error  Unexpected constant condition  no-constant-condition
  249:12 error  Unexpected constant condition  no-constant-condition

hitcon-online/extensions/jitsi/client.mjs
  41:21  error  'JitsiMeetExternalAPI' is not defined  no-undef

hitcon-online/common/extlib/client-extension-helper.mjs
  127:49 error  'methodNam' is not defined             no-undef

hitcon-online/common/extlib/extension-helper-standalone.mjs
  117:5  error  'data' is not defined                  no-undef
  118:12 error  'data' is not defined                  no-undef

hitcon-online/extensions/cellsettest/client.mjs
  19:19  error  'io' is not defined                    no-undef

hitcon-online/extensions/chat/client.mjs
  58:12  error  '$' is not defined                     no-undef
  70:20  error  '$' is not defined                     no-undef
  71:15  error  '$' is not defined                     no-undef

✖ 14 problems (14 errors, 0 warnings)
```

---

加上 `google` 規則集並排除 `valid-jsdoc`、`require-jsdoc` 兩條規則後：
```diff
-✖ 104 problems (14 errors, 90 warnings)
+✖ 209 problems (155 errors, 54 warnings)
+  91 errors and 0 warnings potentially fixable with the `--fix` option.
```
經過 auto fix 後比原先多了這幾條錯誤：
```
32 error    Expected an error object to be thrown  no-throw-literal
 9 warning  This line has a length of 1XX. Maximum allowed is 120  max-len
 7 warning  Identifier 'XXX' is not in camel case  camelcase
 7 error    A function with a name starting with an uppercase letter should only be used as a constructor  new-cap
 6 warning  'assert' is defined but never used  no-unused-vars
 6 error    Unexpected constant condition  no-constant-condition
 6 error    'XXX' is not defined  no-undef
 4 error    The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype  guard-for-in
 3 warning  'XXX' is assigned a value but never used  no-unused-vars
 3 error    A constructor name should not start with a lowercase letter  new-cap
 1 error    Expected the Promise rejection reason to be an Error  prefer-promise-reject-errors
```

**Comment by ktpss95112** [comment]
Every error that emerged after `eslint:recommended` seems solvable.

`hitcon-online/common/extlib/client-extension-manager.mjs`
All the error are caused by bad code. They should be fixed by `typeof abc !== 'def'`.

`hitcon-online/common/rpc-directory/data-store.mjs`
The `while (true)` error can be suppressed by using `for (;;)` as [this stackoverflow](https://stackoverflow.com/questions/37466508/why-does-eslint-trigger-lint-errors-on-whiletrue-using-fibers) suggests.

`hitcon-online/extensions/jitsi/client.mjs`
This may be caused by a broken dependency. I think adding `JitsiMeetExternalAPI = null;` and some comments at the beginning of the file is a reasonable workaround for now.

`hitcon-online/common/extlib/client-extension-helper.mjs`
Just a easily-fixable typo.

`hitcon-online/common/extlib/extension-helper-standalone.mjs`
`return await this.dir.storage.loadData(`ext_${this.name}`);` should be sufficient to solve the error.

`hitcon-online/extensions/cellsettest/client.mjs`
This is solved in #94.

`hitcon-online/extensions/chat/client.mjs`
[This stackoverflow](https://stackoverflow.com/questions/39510736/eslint-dollar-is-not-defined-no-undef) provides the solution.

File: common/extlib/client-extension-manager.mjs (hunk 2 above)
Comment (ktpss95112) [reply common/extlib/client-extension-manager.mjs:120]: It seems that `no-constant-condition` catches these bugs.

**Comment by Sea-n** [comment]
The git hook only warns and blocks commit when ESLint errors are detected.
Because after the script auto-fix the file, the new changes are not in the "ready to commit" stage.
If we just `git add` the file, it may cause unintended results for partial commit.

**Comment by Sea-n** [comment]
ESLint config file for GitHub Actions CI is in `hitcon-online/.github/eslintrc.yml`.
The `hitcon-online/.eslintrc.yml` remains the same.

**Review by john0312 — APPROVED** [review]

File: extensions/items/common/client.mjs [hunk 3]
```diff
@@ -33,16 +33,16 @@ class Client {
    * toPlayerID: string;
    */
   async giveItem(toPlayerID, itemName, amount) {
-    if (!itemName in this.itemInfo) {
-      console.error('Item does not exist');
+    if (!(itemName in this.itemInfo)) {
+      console.log("Item does not exist");
```
Comment (john0312) [inline extensions/items/common/client.mjs:37]: Is this intentional?

Comment (john0312) [reply extensions/items/common/client.mjs:37]: Note: There are many instances of replacing error() with log() below as well.

File: run/start-all.mjs [hunk 4]
```diff
@@ -52,23 +52,23 @@ async function main() {
   redisClient.quit();
 
   /* Start asset server */
-  const assetServer = fork('../services/assets/asset-server-launcher.mjs', { cwd: '.' });
```
Comment (john0312) [inline run/start-all.mjs:55]: Is this intentional?

File: run/start-all.mjs [hunk 5]
```diff
@@ -52,23 +52,23 @@ async function main() {
...
+  const assetServer = fork('../services/main/asset-server.mjs', {cwd: '.'});
 
   /* Start gateway service */
   const gatewayServers = {};
   const enabledGatewayServers = config.get('gatewayServers');
-  for(const serverName in enabledGatewayServers){
-    gatewayServers[serverName] = fork('../services/gateway/gateway-server.mjs', ['--service-name', serverName], { cwd: '.' });
```
Comment (john0312) [inline run/start-all.mjs:61]: Is this intentional?

File: run/start-all.mjs [hunk 6]
```diff
@@ -80,8 +80,9 @@ async function main() {
   /* Start standalone extension services */
   const extServices = {};
   const enabledExtStandalone = config.get('ext.standalone');
-  for(const ext in enabledExtStandalone){
-    extServices[ext] = fork('../services/standalone/standalone-extension.mjs', ['--ext', ext], { cwd: '.' });
+  for (const ext in enabledExtStandalone) {
+    if (config.get('ext.enabled').includes(ext)) continue;
```
Comment (john0312) [inline run/start-all.mjs:84]: Is this intentional?

File: run/start-all.mjs [hunk 7]
```diff
@@ -93,31 +94,33 @@ async function main() {
...
-    for(const serverName in enabledGatewayServers){
+    for (const serverName in enabledGatewayServers) {
       try {
         gatewayServers[serverName].kill();
       } catch {};
     }
 
-    for(const ext in enabledExtStandalone){
+    for (const ext in enabledExtStandalone) {
+      if (config.get('ext.enabled').includes(ext)) continue;
```
Comment (john0312) [inline run/start-all.mjs:104]: Is this intentional?

File: run/start-all.mjs [hunk 8]
```diff
@@ -93,31 +94,33 @@ async function main() {
...
-    gatewayServers[serverName].on('close', (err) => { handler(serverName + ' service closed.', err) });
+  for (const serverName in enabledGatewayServers) {
+    gatewayServers[serverName].on('error', (err) => { handler(serverName + ' service error.', err); });
+    gatewayServers[serverName].on('close', (err) => { handler(serverName + ' service closed.', err); });
   }
-  for(const ext in enabledExtStandalone){
-    extServices[ext].on('error', (err) => { handler(ext + ' service error.', err) });
-    extServices[ext].on('close', (err) => { handler(ext + ' service closed.', err) });
+  for (const ext in enabledExtStandalone) {
+    if (config.get('ext.enabled').includes(ext)) continue;
```
Comment (john0312) [inline run/start-all.mjs:121]: Is this intentional?

**Comment by Sea-n** [comment]
Except of `eslint --fix`, I only changed `for (!item in arr)` to `for (!(item in arr))`.
Other syntax changes are unintended.

Sorry for not checking the changes before PR :sweat:

**Pull request merged**